### PR TITLE
Iterable Channels

### PR DESCRIPTION
**Pull request by trowski**
Allows iteration of `Channel` objects using `foreach`. This also forbids `null` to be sent on a channel, because `receive()` returns null once the channel closes, similar to `read()` returning null when a stream closes on `ReadableStream`.

These changes provide what I think is an intuitive and easy-to-use interface. One coroutine can consume values from another coroutine without the need for try/catch or sending/checking for a stop marker value.

```php
$values = [1, 2, 3];
[$left, $right] = createChannelPair();

async(function () use ($right, $values): void {
    foreach ($values as $value) {
        $right->send($value);
    }

    $right->close();
});

foreach ($left as $received) {
    self::assertSame(\array_shift($values), $received);
}
```

### CODE REVIEW

**Comment by kelunik** [comment]
It's pretty late in the release cycle to forbid sending null IMO. We can also keep the current behavior and still make it iterable, no?

**Comment by trowski** [comment]
> It's pretty late in the release cycle to forbid sending null IMO.

Probably yes. The `Channel` implementation in this library was still forbidding sending null, though I doubt it's seen much use. I'm sure we discussed sending null sometime in the last few years and decided to allow it. It is nice to avoid arbitrary limitations and have parity with concurrent iterators. I've just not used it myself outside of sending a stop/close signal to avoid the `ChannelException` from `receive()`.

> We can also keep the current behavior and still make it iterable, no?

Yes. The iterator will have to be implementation specific to avoid creating an exception, and `receive()` will still throw on a normal close. For example, [here](https://github.com/amphp/cluster/blob/v2/src/Cluster.php#L165) I needed to catch the exception. Since a cancellation is used with `receive()`, this PR will not change that code.

I updated the PR to allow sending null again on a channel. I've also updated `StreamChannel` in byte-stream in my local branch to implement `IteratorAggregate` to be sure there were no unforeseen issues in that class. The only other channel implementations we ship are a wrapper around a ConcurrentIterator (similar to `ConcurrentIteratorChannel`) or a decorator.

File: src/Internal/ConcurrentIteratorChannel.php [hunk 1]
```diff
@@ -71,24 +73,42 @@ public function onClose(\Closure $onClose): void
...
+            0,
+            $exception ?? null,
+        );
     }
 
     public function send(mixed $data): void
     {
-        if ($data === null) {
-            throw new ChannelException("Cannot send null on a channel");
-        }
```
Comment (kelunik) [inline src/Internal/ConcurrentIteratorChannel.php:86]: Why did you remove this restriction now?

File: src/Channel.php [hunk 2]
```diff
@@ -9,24 +9,26 @@
...
  */
-interface Channel extends Closable
+interface Channel extends Closable, \Traversable
 {
     /**
      * @param Cancellation|null $cancellation Cancels waiting for the next value. Note the next value is not discarded
      * if the operation is cancelled, rather it will be returned from the next call to this method.
      *
-     * @return TReceive Data received.
+     * @return TReceive|null Data received or {@code null} if the channel has closed.
```
Comment (kelunik) [inline src/Channel.php]: We allow sending `null` now but also use it as termination signal?

Comment (trowski) [reply src/Channel.php]: Oops, forgot to revert this annotation back. Fixed.

File: src/Internal/ConcurrentIteratorChannel.php (hunk 1 above)
Comment (trowski) [reply src/Internal/ConcurrentIteratorChannel.php:86]: I guess this was missed when we allowed null on channels.

File: src/Channel.php [hunk 3]
```diff
@@ -9,24 +9,26 @@
...
 {
     /**
      * @param Cancellation|null $cancellation Cancels waiting for the next value. Note the next value is not discarded
      * if the operation is cancelled, rather it will be returned from the next call to this method.
      *
-     * @return TReceive Data received.
+     * @return TReceive
      *
-     * @throws ChannelException If receiving from the channel fails or the channel closed.
+     * @throws ChannelException If receiving from the channel fails.
```
Comment (kelunik) [inline src/Channel.php:25]: This should stay then?

File: src/Channel.php [hunk 4]
```diff
@@ -9,24 +9,26 @@
...
      * @throws SerializationException If the underlying transport mechanism uses serialization and fails.
      */
     public function receive(?Cancellation $cancellation = null): mixed;
 
     /**
-     * @param TSend $data
+     * @param TSend $data Any value other than {@code null}.
```
Comment (kelunik) [inline src/Channel.php:31]: This restriction has been removed, no?

File: src/Internal/ConcurrentIteratorChannel.php [hunk 5]
```diff
@@ -71,24 +73,42 @@ public function onClose(\Closure $onClose): void
...
+        try {
+            foreach ($this->receive as $value) {
+                yield $value;
+            }
+        } catch (DisposedException $exception) {
+            throw new ChannelException(
+                "The channel closed while waiting to receive the next value",
+                0,
+                $exception,
+            );
```
Comment (kelunik) [inline src/Internal/ConcurrentIteratorChannel.php:111]: Why does this still throw instead of simply returning?

**Comment by trowski** [comment]
After discussing with @kelunik, I'm closing this as this. Without a pre-defined stop signal, implementing iterator seems awkward and misplaced on this interface.
